File: go/016.go
func threeSumClosest(nums []int, target int) int {
    sort.Ints(nums)
    size, ret, closest := len(nums), 0, math.MaxInt32
    for i := 0; i < size-2; i++ {
        for j, k := i+1, size-1; j < k; {
            temp := nums[i] + nums[j] + nums[k]
            if temp == target {
                return target
            } else if temp < target {
                j++
            } else {
                k--
            }
            newDiff := abs(temp - target)
            if newDiff < closest {
                closest = newDiff
                ret = temp
            }
        }
    }
    return ret
}

func abs(a int) int {
    if a < 0 {
        return -a
    }
    return a
}

File: cpp/016.cpp
class Solution {
public:
    int threeSumClosest(vector<int>& nums, int target) {
        sort(nums.begin(), nums.end());
        int size = nums.size(), closest = INT_MAX, ret = 0;
        for (int i = 0; i < size - 2; ++i) {
            for (int j = i + 1, k = size - 1; j < k; ) {
                int temp = abs(nums[i]+nums[j]+nums[k]-target);
                if (temp == 0)
                    return target;
                if (temp < closest) {
                    closest = temp;
                    ret = nums[i]+nums[j]+nums[k];
                } 
                if (nums[i]+nums[j]+nums[k] < target)
                    ++j;
                else --k;
            }
        }
        return ret;
    }
};
